voyager-navigator/src/commonMain/kotlin/cafe/adriel/voyager/navigator/internal/NavigatorBackHandler.kt

```diff
@@ -12,14 +12,21 @@ internal fun NavigatorBackHandler(
     navigator: Navigator,
     onBackPressed: OnBackPressed
 ) {
-    if (onBackPressed != null) {
+    if (onBackPressed == null) {
         BackHandler(
-            enabled = navigator.canPop || navigator.parent?.canPop ?: false,
+            enabled = navigator.size > 1,
+            onBack = navigator::popRecursively
+        )
+    } else {
+        // `navigator.size == 1` covers onBackPressed = { false } and empty stack
+        // because `navigator.canPop` always returns `false` when stack min size is 1
+        BackHandler(
+            enabled = (navigator.size == 1 && !onBackPressed(navigator.lastItem)) ||
```

Comment:
I don't think we should use the `onBackPressed` function here. The assumption is that the `onBackPressed` will be called when is actually poping the navigation.